design/004-external-data-handling/004-external-data-handling.md

```diff
@@ -0,0 +1,232 @@
+that is outside of the repository. For example
+
+```shell
+renku dataset add --external ../../path/to/my/data
+```
+
+The path to the data is stored in the dataset's metadata; the pointer can also
+be updated in case that the data changes and the hash needs to be recomputed,
+for the purposes of reproducibility. Of course, the obvious drawback here is
+that the data location is arbitrary - if the project is brought to another
+environment/machine/cluster, it's fairly likely that the data path will be
+incorrect and therefore all recorded workflows will fail.
+
+#### S3
+
+Using S3 in Renku is currently limited to interactive sessions. When launching a
+session, users can specify any S3-compatible URL together with credentials, if
+necessary. The S3 bucket is then mounted in their interactive session.
+Currently, no metadata about these mounts is recorded. The data could, in
+principle, be added to a renku dataset with the "external" mechanism described
+above, but no metadata about its actual origins (i.e. an S3 bucket) would be
+recorded.
+
+#### Upcoming
+
+We have plans to include other external data providers, such as NFS shares,
+openBIS etc. From the user's perspective, they all will generally work in the
+same way as S3; data will be mounted to a local directory outside the repository
+and made accessible as network-mounted storage.
+
+## Design Detail
+
+Generically, we should strive to remove the separation between "hosted" Renku
+(i.e. what I can do if I run an interactive session on renkulab.io) and
+"offline" Renku (what I can do with the CLI anywhere).
+
+Ideally we could once again follow the plug-in paradigm in the CLI, like we
+do for other functionality.
+
+### Provider interface
+
+tbd
+
+### Commands
+
+#### Adding external data:
+
+```
+renku dataset add s3://server/path/to/bucket
+Access Key: 1234
+Secret Key: abcd
+Store credentials? (y/N): y
+Credentials stored in ~/.renku/renku.ini
+```
+
+`s3://` is used here but we should also support at least `nfs://`. Renku should
+ask for credentials in a way that is appropriate for the type of resource that
+is being added.
+
+#### Getting information about the dataset:
+
+```
+renku dataset show very-large-data
+
+Name: very-large-data
+Created: 2022-03-30 15:55:59+02:00
+Creator(s): Rok Roškar <rok.roskar@sdsc.ethz.ch>
+Title: very-large-data
+Description: Data that is very large. Indeed.
+External Sources:
+    - type: s3
+      source: s3://server/path/to/bucket
+      location: None
+```
+
+#### Staging external data:
+
+External data has to somehow be brought into the system where Renku is running.
+This seems to be a "staging" operation, but we could also call it something
+else. "Mounting" is another option but seems too linux-specific.
+
+```
+renku dataset stage very-large-data
+```
+
+This command uses Fuse when necessary to mount data locally. Once it's staged,
+it can be found locally and used in workflows:
```

Comment:
Well since cwl (and toil?) copy files to a temporary directory for execution, we would also need to handle that properly, linking instead of copying or submitting absolute paths.

Reply:
are there built-in mechanisms in toil and/or cwl to do this automatically? @vigsterkr maybe you know?

Reply:
indeed there is a `--linkImports` option that will do a hard-link. But my guess is that this won't really work with fuse-mounted fs.